src/main/cpp/src/cast_string_to_timestamp.cu

```diff
@@ -0,0 +1,876 @@
+ * @param ptr The input string
+ * @param[out] pos the pointer to the string when parsing
+ * @param end_pos the end pointer of the string
+ * @param[out] v the parsed value
+ * @param min_digits the minimum digits to parse, if not enough, return false
+ * @param max_digits the maximum digits to parse, if exceeds, return false
+ * @return true if the input string is valid, false otherwise
+ */
+__device__ bool parse_int(
+  char const* const ptr, int& pos, int const end_pos, int& v, int min_digits, int max_digits)
+{
+  v          = 0;
+  int digits = 0;
+  while (pos < end_pos) {
+    int const parsed_value = static_cast<int32_t>(ptr[pos] - '0');
+    if (parsed_value >= 0 && parsed_value <= 9) {
+      if (++digits > max_digits) { return false; }
+      v = v * 10 + parsed_value;
+    } else {
+      break;
+    }
+    pos++;
+  }
+
+  return digits >= min_digits;
+}
+
+/**
+ * Parse a string to an integer for microseconds, microseconds has 0-6 digits.
+ * Spark store timestamp in microsecond, the nanosecond part is discard.
+ * If there are 6+ digits, only use the first 6 digits, the rest digits are
+ * ignored/truncated.
+ * @param ptr the current pointer to the string
+ * @param[out] pos the pointer to the string when parsing
+ * @param end_pos the end pointer of the string
+ * @param[out] v the parsed value
+ */
+__device__ void parse_microseconds(char const* const ptr, int& pos, int const end_pos, int& v)
+{
+  v          = 0;
+  int digits = 0;
+  while (pos < end_pos) {
+    int const parsed_value = static_cast<int32_t>(ptr[pos] - '0');
+    if (parsed_value >= 0 && parsed_value <= 9) {
+      if (++digits <= 6) {
+        v = v * 10 + parsed_value;
+      } else {
+        // ignore/truncate the rest digits
+      }
+    } else {
+      break;
+    }
+    pos++;
+  }
+}
+
+__device__ bool eof(int pos, int end_pos) { return end_pos - pos <= 0; }
+
+__device__ bool parse_char(char const* const ptr, int& pos, char c) { return ptr[pos++] == c; }
+
+__device__ bool try_parse_char(char const* const ptr, int& pos, char const c)
+{
+  if (ptr[pos] == c) {
+    ++pos;
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Try to parse `max_digits` consecutive digits into a int value.
+ * Exits on the first non-digit character.
+ * @param ptr The input string
+ * @param[out] pos the pointer to the string when parsing
+ * @param end_pos the end pointer of the string
+ * @param[out] v the parsed value
+ * @param max_digits the maximum digits to parse
+ * @returns the number of digits parsed.
+ */
+__device__ int parse_digits(
+  char const* const ptr, int& pos, int const end_pos, int& v, int const max_digits)
+{
+  v          = 0;
+  int digits = 0;
+  while (pos < end_pos) {
+    int const parsed_value = static_cast<int32_t>(ptr[pos] - '0');
+    if (parsed_value >= 0 && parsed_value <= 9) {
+      v = v * 10 + parsed_value;
+      ++pos;
+      if (++digits == max_digits) { break; }
+    } else {
+      // meets non-digit
```

Comment:
So there is no error report needed for parsing invalid digits?

Reply:
Yes, no error report.
The `parse_digits` is similar to `parse_int`, the diff is that `parse_digits` returns the parsed number of parsed digits.

In `parse_tz_from_sign`, it needs to check the parsed digits number:
```cpp
        if (!(m_digits == 2 && (s_digits == 0 || s_digits == 2) && eof(pos, end_pos))) {
          // not: [+-]h[h][mm] or [+-]h[h][mm][ss]
          return make_invalid_tz();
        }
```